lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,278 @@
+    #          containing error message, and an HTTP code
+    #
+    # Examples:
+    #   Responses if success:
+    #     IPv4: {"code": 200, "data": "100.55.55.3"}
+    #     IPv6: {"code": 200, "data": "2001:db8:abcd:12::1"}
+    #   Response if missing required params:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet does not exist:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"code": 404, "error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse get("/subnet/#{subnet}/next_ip?mac=#{mac}&group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['data']
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from External IPAM."), subnet)
+    end
+
+    # Adds an IP address to the specified subnet in External IPAM. This will reserve the IP in the
+    # External IPAM database. If group is specified, the IP will be added to the subnet within the
+    # given group.
+    #
+    # Params: 1. ip:               IP address to be added
+    #         2. subnet:           The IPv4 or IPv6 subnet CIDR. (Examples: IPv4 - "100.10.10.0/24",
+    #                              IPv6 - "2001:db8:abcd:12::/124")
+    #         3. group(optional):  The name of the External IPAM group containing the subnet.
+    #
+    # Returns: true if IP was added successfully to External IPAM, otherwise false
+    #
+    # Responses from Proxy plugin:
+    #   Response if success:
+    #     {"code": 200}
+    #   Response if IP already reserved:
+    #     {"code": 409, "error": "IP address already exists"}
+    #   Response if subnet error:
+    #     {"code": 404, "error": "The specified subnet does not exist in External IPAM."}
+    #   Response if missing required params:
+    #     {"code": 400, "error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if can't connect to External IPAM server
+    #     {"code": 500, "error": "Unable to connect to External IPAM server"}
+    def add_ip_to_subnet(ip, subnet, group = "")
+      raise "subnet cannot be nil" if subnet.nil?
+      response = parse post({}, "/subnet/#{subnet}/#{ip}?group=#{URI.escape(group.to_s)}")
+      raise ::Foreman::Exception.new(N_(response['error'])) if response['error']
+      response['code'] == 200
+    rescue => e
```

Comment:
Nitpick: You could actually override the `parse` method in this class, call `super` and add your own error handling code (the `if response['error']` line). Then all the methods using `parse` call would become much cleaner.